### ISSUE TITLE
Add Support for Tags

### ISSUE THREAD

**Issue by Rperry2174** [pyroscope-io/pyroscope#162]
Many have requested support for tags. While we built the backend in a way where it should be able to support tags fairly easily we just need to figure out the best way for labeling things (i.e. by IP, Kubernetes Pod, etc)

**Comment by petethepig** [comment]
TODO:
* [ ] change dictionary to be per app, not per segment
* [ ] run some tests, see how bad performance gets with a lot of tags. (how does it perform with a million tags?)
* [ ] find a way to address performance issues (sampling?)
* [ ] make sure we expose all the endpoints needed to make tags work on the frontend
* [ ] write something in docs
* [ ] 


Frontend TODO:
* [ ] add support for tags to grafana plugin

**Comment by kolesnikovae** [comment]
I've prepared [a basic benchmark](https://github.com/pyroscope-io/pyroscope/blob/tags-benchmarks/pkg/storage/cardinality_test.go) for throughput. As a reference I measured `Put` and `Get` throughput with no tags for a single application:
```
BenchmarkTags/no_tags_(0)
BenchmarkTags/no_tags_(0)/put
BenchmarkTags/no_tags_(0)/put-4     	                    6270	    224669 ns/op
BenchmarkTags/no_tags_(0)/get
BenchmarkTags/no_tags_(0)/get-4     	                   33836	     51987 ns/op
```
Here 6270 and 33836 is `Put` and `Get` op/s, respectively. The database is populated with 25k samples for `Get` benchmark.

----

The next measurement is for samples with 5 tags, 12.5k unique combinations. For `Get` benchmark, cardinality of the tag is specified in parenthesis.

```
BenchmarkTags/5_tags_(12.5k)
BenchmarkTags/5_tags_(12.5k)/put
BenchmarkTags/5_tags_(12.5k)/put-4  	                    1471	    804129 ns/op
BenchmarkTags/5_tags_(12.5k)/get_(app)
BenchmarkTags/5_tags_(12.5k)/get_(app)-4         	     528	   2394778 ns/op
BenchmarkTags/5_tags_(12.5k)/get_(card_5)
BenchmarkTags/5_tags_(12.5k)/get_(card_5)-4      	    2480	    429782 ns/op
BenchmarkTags/5_tags_(12.5k)/get_(card_25)
BenchmarkTags/5_tags_(12.5k)/get_(card_25)-4     	    9458	    121429 ns/op
BenchmarkTags/5_tags_(12.5k)/get_(card_100)
BenchmarkTags/5_tags_(12.5k)/get_(card_100)-4    	   23578	     45241 ns/op
```

----

The last measurement has been done for samples with 5 tags, cardinality 25k (every of 25k objects in db is unique).
```
BenchmarkTags/5_tags_(inf)
BenchmarkTags/5_tags_(inf)/put
BenchmarkTags/5_tags_(inf)/put-4                 	    1912	    723047 ns/op
BenchmarkTags/5_tags_(inf)/get_(app)
BenchmarkTags/5_tags_(inf)/get_(app)-4           	       1	2975326978 ns/op
BenchmarkTags/5_tags_(inf)/get_(card_5)
BenchmarkTags/5_tags_(inf)/get_(card_5)-4        	       6	 227955529 ns/op
BenchmarkTags/5_tags_(inf)/get_(card_25)
BenchmarkTags/5_tags_(inf)/get_(card_25)-4       	      43	  38850761 ns/op
BenchmarkTags/5_tags_(inf)/get_(card_100)
BenchmarkTags/5_tags_(inf)/get_(card_100)-4      	     100	  11801381 ns/op
BenchmarkTags/5_tags_(inf)/get_(card_1000000)
BenchmarkTags/5_tags_(inf)/get_(card_1000000)-4  	    1093	   1525207 ns/op
```
----

Pretty obvious conclusions:
1. `Put` performance is inversely proportional to the number of tags. The relation is linear.
2. `Get` performance is inversely proportional to the tag cardinality: the lowest performance is observed when no tags specified (maximal cardinality).

Although with moderate cardinality storage is still operational, perhaps we should revise the current design. I started working on a proposal for ingestion aggregation - instead of merging trees at querying, ones could be merged at ingestion.

**Comment by Rperry2174** [comment]
closed in #296
